keel-actuator/src/main/kotlin/com/netflix/spinnaker/keel/actuation/EnvironmentPromotionChecker.kt

```diff
@@ -12,10 +16,21 @@ import org.springframework.stereotype.Component
 @Component
 class EnvironmentPromotionChecker(
   private val artifactRepository: ArtifactRepository,
+  private val deliveryConfigRepository: DeliveryConfigRepository,
   private val constraints: List<ConstraintEvaluator<*>>,
   private val publisher: ApplicationEventPublisher
 ) {
+  private val statefulEvaluators: List<ConstraintEvaluator<*>> = constraints
+    .filterIsInstance<StatefulConstraintEvaluator<*>>()
+  private val statelessEvaluators: List<ConstraintEvaluator<*>> = constraints
+    .filter { !statefulEvaluators.contains(it) }
 
+  /**
+   * TODO (Critical): The EC2 ClusterSpec enables artifact filtering by ArtifactStatus. This is
+   *  currently handled in ImageResolver after artifacts are "approved" for an environment.
+   *  Instead, this needs to happen here or earlier. Otherwise we may deploy canary clusters or
+   *  request manual judgements for artifacts that should never be allowed in a given environemnt.
+   */
```

Comment:
We need to refactor how ArtifactStatus constraints are defined and evaluated. This likely means establishing a rule of only one ArtifactStatus restriction per environment and may include making it or a top-level `Environment` property.